### PR TITLE
feat(queue): add queue capabilty to sdk

### PR DESCRIPTION
**Pull request by Yangqing**
manually tested: `python leptonai/tests/test_queue.py` on latest.

### CODE REVIEW

File: leptonai/queue.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        :param Connection conn: the connection to use. If None, use the default workspace connection.
+        """
+        self._conn = conn if conn else workspace_api.current_connection()
+        res = queue_api.list_queue(self._conn)
+        if not res.ok:
+            raise RuntimeError(
+                "Failed to access queue server. Error:"
+                f" {res.status_code} {res.content}."
+            )
+        existing_queues = [s["name"] for s in res.json()]
```
Comment (bddppq) [inline leptonai/queue.py:105]: maybe better to use `existing_queue = Queue.list_queue(self._conn)`

File: leptonai/queue.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+                # wait for the queue to be ready
+                start = time.time()
+                while (
+                    not self.ready()
+                    and time.time() - start < _lepton_readiness_timeout_
+                ):
+                    logger.trace(f"Queue {name} is not ready yet. Waiting...")
+                    time.sleep(_lepton_readiness_wait_time_)
+                if not self.ready():
```
Comment (bddppq) [inline leptonai/queue.py:133]: minor optimization: I think we can save one network call here:
```
while time.time() - start < _lepton_readiness_timeout_:
    if self.ready():
        break
    time.sleep(_lepton_readiness_wait_time_)
else:
    raise RuntimeError(...)
```

File: leptonai/queue.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                raise ValueError(
+                    f"Queue {name} does not exist, and you have specified"
+                    " create_if_not_exists=False."
+                )
+
+    def ready(self) -> bool:
+        existing_queues = Queue.list_queue(self._conn)
+        return self._queue in existing_queues
+
+    async def async_wait(self):
```
Comment (bddppq) [inline leptonai/queue.py:148]: Do we need this method? I think a queue instance is always ready to be use after creation (we wait for ready in `__init__`)

**Review by bddppq — APPROVED** [review]
Overall LGTM

File: leptonai/api/queue.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+def receive(conn: Connection, name: str):
+    """
+    Receives a message from a queue.
+    """
+    response = conn.get(f"/queue/{name}/messages")
+    return response
+
+
+def send(conn: Connection, name: str, message: Union[str, bytes]):
```
Comment (bddppq) [inline leptonai/api/queue.py:39]: I think message can only be of type `str`? Posting bytes in json will raise error IIRC

File: leptonai/queue.py (hunk 3 above)
Comment (Yangqing) [reply leptonai/queue.py:148]: my fault, I was refactoring it from the kv impl and did not include the wait argument in `__init__`

File: leptonai/api/queue.py (hunk 4 above)
Comment (Yangqing) [reply leptonai/api/queue.py:39]: yeah... will fox.

File: leptonai/queue.py (hunk 1 above)
Comment (Yangqing) [reply leptonai/queue.py:105]: will do

File: leptonai/queue.py (hunk 2 above)
Comment (Yangqing) [reply leptonai/queue.py:133]: sounds good

File: leptonai/api/queue.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    """
+    Put a key-value pair in the queue.
```
[the remark this answers is not in the document]
Comment (Yangqing) [reply leptonai/api/queue.py:41]: done

**Pull request merged**